include/arch/x86/processor.h

```diff
@@ -201,15 +201,21 @@
 #include <stdbool.h>
 #include <stdint.h>
 
+#if defined(__x86_64__)
+typedef uint64_t x86_reg_t;
+#elif defined(__i386__)
+typedef uint32_t x86_reg_t;
+#endif
+
 union x86_ex_error_code {
-    uint32_t error_code;
+    x86_reg_t error_code;
     struct __packed {
-        unsigned int E : 1, TLB : 2, index : 13, rsvd_sel : 16;
+        x86_reg_t E : 1, TLB : 2, index : 13;
     };
```

Comment:
Uhm, I believe you need to specify the padding bits still which gets a little awkward when we want to support 32-/64-bit types. Otherwise it's uncertain in what order and where those bits will be allocated.

[gcc's documentation](https://gcc.gnu.org/onlinedocs/gcc/Structures-unions-enumerations-and-bit-fields-implementation.html) is a little scarce about it:

> * The order of allocation of bit-fields within a unit (C90 6.5.2.1, C99 and C11 6.7.2.1).
> 
> Determined by ABI.

I, myself, can't read anything out of this, beside, "if it breaks, you're free to keep all pieces :P".

Maybe we should make it portable and simply use masks to access the individual bits?

Reply:
Yes, will fix it. But I really do not want to use the masks :-).